File: 2023/kotlin/src/Day07.kt
private enum class HandType {
    HIGH_CARD, ONE_PAIR, TWO_PAIR, THREE_OF_A_KIND, FULL_HOUSE, FOUR_OF_A_KIND, FIVE_OF_A_KIND
}

private open class Hand(val cards: String, open val bid: Int): Comparable<Hand> {
    protected open val type: HandType
        get() {
            val frequencies = cards.groupingBy { it }.eachCount().values.sortedDescending()

            return when (frequencies) {
                listOf(5) -> HandType.FIVE_OF_A_KIND
                listOf(4, 1) -> HandType.FOUR_OF_A_KIND
                listOf(3, 2) -> HandType.FULL_HOUSE
                listOf(3, 1, 1) -> HandType.THREE_OF_A_KIND
                listOf(2, 2, 1) -> HandType.TWO_PAIR
                listOf(2, 1, 1, 1) -> HandType.ONE_PAIR
                else -> HandType.HIGH_CARD
            }
        }

    protected open val cardRanks = "23456789TJQKA"

    override fun compareTo(other: Hand): Int {
        val comparison = this.type.compareTo(other.type)

        if (comparison != 0) {
            return comparison
        }

        cards.zip(other.cards).forEach { (a, b) ->
            val cardComparison = cardRanks.indexOf(a) - cardRanks.indexOf(b)
            if (cardComparison != 0) {
                return cardComparison
            }
        }

        return 0
    }

    companion object {
        fun of(line: String): Hand {
            val parts = line.split(" ")
            return Hand(parts.first(), parts.last().toInt())
        }
    }
}

private class WildJokerHand(cards: String, bid: Int): Hand(cards, bid) {
    override val type: HandType
        get() {
            val cardsWithoutJokers = cards.filterNot { it == 'J' }

            if (cardsWithoutJokers.isEmpty()) {
                return HandType.FIVE_OF_A_KIND
            }

            val frequencies = cardsWithoutJokers.groupingBy { it }.eachCount().values.sortedDescending().toMutableList()
            frequencies[0] += cards.length - cardsWithoutJokers.length

            return when (frequencies) {
                listOf(5) -> HandType.FIVE_OF_A_KIND
                listOf(4, 1) -> HandType.FOUR_OF_A_KIND
                listOf(3, 2) -> HandType.FULL_HOUSE
                listOf(3, 1, 1) -> HandType.THREE_OF_A_KIND
                listOf(2, 2, 1) -> HandType.TWO_PAIR
                listOf(2, 1, 1, 1) -> HandType.ONE_PAIR
                else -> HandType.HIGH_CARD
            }
        }

    override val cardRanks = "J23456789TQKA"

    companion object {
        fun of(line: String): WildJokerHand {
            val parts = line.split(" ")
            return WildJokerHand(parts.first(), parts.last().toInt())
        }
    }
}

private fun part1(input: String) =
    input
        .lines()
        .map { Hand.of(it) }
        .sorted()
        .mapIndexed { rank, hand -> (rank + 1) * hand.bid }
        .sum()

private fun part2(input: String) =
    input
        .lines()
        .map { WildJokerHand.of(it) }
        .sorted()
        .mapIndexed { rank, hand -> (rank + 1) * hand.bid }
        .sum()

fun main() {
    val input = readInput("Day07")
    part1(input).println()
    part2(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day07.swift
extension String {
  fileprivate func counts() -> [Character: Int] {
    var frequencies = [Character: Int]()

    self.forEach {
      frequencies[$0, default: 0] += 1
    }

    return frequencies
  }
}

struct Day07: AdventDay {
  var input: String

  enum HandType: Int {
    case HIGH_CARD, ONE_PAIR, TWO_PAIR, THREE_OF_A_KIND, FULL_HOUSE, FOUR_OF_A_KIND, FIVE_OF_A_KIND
  }

  class Hand: Comparable {
    let cards: String
    let bid: Int

    init(cards: String, bid: Int) {
      self.cards = cards
      self.bid = bid
    }

    func cardRanks() -> String {
      return "23456789TJQKA"
    }

    func type() -> HandType {
      let frequencies = self.cards.counts().values.sorted(by: >)

      switch frequencies {
      case [5]:
        return HandType.FIVE_OF_A_KIND
      case [4, 1]:
        return HandType.FOUR_OF_A_KIND
      case [3, 2]:
        return HandType.FULL_HOUSE
      case [3, 1, 1]:
        return HandType.THREE_OF_A_KIND
      case [2, 2, 1]:
        return HandType.TWO_PAIR
      case [2, 1, 1, 1]:
        return HandType.ONE_PAIR
      default:
        return HandType.HIGH_CARD
      }
    }

    static func < (lhs: Hand, rhs: Hand) -> Bool {
      if lhs.type().rawValue == rhs.type().rawValue {
        for pair in zip(lhs.cards, rhs.cards) {
          if lhs.cardRanks().firstIndex(of: pair.0)! != rhs.cardRanks().firstIndex(of: pair.1)! {
            return lhs.cardRanks().firstIndex(of: pair.0)! < lhs.cardRanks().firstIndex(of: pair.1)!
          }
        }

        return false
      }

      return lhs.type().rawValue < rhs.type().rawValue
    }

    static func == (lhs: Hand, rhs: Hand) -> Bool {
      lhs.cards == rhs.cards
    }
  }

  class WildJokerHand: Hand {
    override func cardRanks() -> String {
      return "J23456789TQKA"
    }

    override func type() -> HandType {
      let cardsWithoutJokers = self.cards.filter { $0 != "J" }

      if cardsWithoutJokers.isEmpty {
        return HandType.FIVE_OF_A_KIND
      }

      var frequencies = cardsWithoutJokers.counts().values.sorted(by: >)
      frequencies[0] += self.cards.count - cardsWithoutJokers.count

      switch frequencies {
      case [5]:
        return HandType.FIVE_OF_A_KIND
      case [4, 1]:
        return HandType.FOUR_OF_A_KIND
      case [3, 2]:
        return HandType.FULL_HOUSE
      case [3, 1, 1]:
        return HandType.THREE_OF_A_KIND
      case [2, 2, 1]:
        return HandType.TWO_PAIR
      case [2, 1, 1, 1]:
        return HandType.ONE_PAIR
      default:
        return HandType.HIGH_CARD
      }
    }
  }

  func winnings(from hands: [Hand]) -> Int {
    zip(1..., hands.sorted())
      .map { rank, hand in
        rank * hand.bid
      }
      .sum()
  }

  func part1() -> Any {
    let hands =
      input
      .lines()
      .map { line in
        let parsed = line.split(separator: " ")
        return Hand(cards: "\(parsed[0])", bid: Int(parsed[1])!)
      }

    return winnings(from: hands)
  }

  func part2() -> Any {
    let hands =
      input
      .lines()
      .map { line in
        let parsed = line.split(separator: " ")
        return WildJokerHand(cards: "\(parsed[0])", bid: Int(parsed[1])!)
      }

    return winnings(from: hands)
  }
}
